src/aiida/orm/fields.py

```diff
@@ -0,0 +1,502 @@
+
+        # only allow an existing fields attribute if has been generated from a subclass
+        current_fields = getattr(cls, 'fields', None)
+        if current_fields is not None and not isinstance(current_fields, QbFields):
+            raise ValueError(f"class '{cls}' already has a `fields` attribute set")
+
+        # Find all fields
+        fields = {}
+        # Note: inspect.getmembers causes loading of AiiDA to fail
+        for key, attr in ((key, attr) for subcls in reversed(cls.__mro__) for key, attr in subcls.__dict__.items()):
+            # __qb_fields__ should be a list of QbField instances
+            if key == '__qb_fields__':
+                assert isinstance(
+                    attr, t.Sequence
+                ), f"class '{cls}' has a '__qb_fields__' attribute, but it is not a sequence"
+                for field in attr:
+                    if not isinstance(field, QbField):
+                        raise ValueError(f"__qb_fields__ attribute of class '{cls}' must be list of QbField instances")
+                    if field.key == 'value':
+                        if dtype := getattr(cls, '_type', None):
+                            value_field = add_field(
+                                'value',
+                                dtype=dtype,
+                                doc='The value of the data',
+                            )
+
+                            fields[field.key] = value_field
+                    else:
+                        fields[field.key] = field
+
+        cls.fields = QbFields({key: fields[key] for key in sorted(fields)})
+
+
+class QbFieldArguments(t.TypedDict):
+    key: str
+    qb_field: t.Optional[str]
+    dtype: t.Optional[t.Any]
+    doc: str
+    is_db_field: bool
+    is_subscriptable: bool
+
+
+def add_field(
+    key: str,
+    qb_field: t.Optional[str] = None,
+    *,
+    dtype: t.Optional[t.Any] = None,
+    doc: str = '',
+    is_db_field: bool = False,
+    is_subscriptable: bool = False,
+) -> QbField:
+    """Add a `dtype`-dependent `QbField` representation of a field.
+
+    :param key: The key of the field on the ORM entity
+    :param qb_field: The name of the field in the QueryBuilder, if not equal to ``key``
+    :param dtype: The data type of the field. If None, the field is of variable type.
+    :param doc: A docstring for the field
+    :param is_db_field: If False, the ``qb_field`` property will prepend "attributes." to field name
+    :param is_subscriptable: If True, a new field can be created by ``field["subkey"]``
+    """
+    kwargs: QbFieldArguments = {
+        'key': key,
+        'qb_field': qb_field,
+        'dtype': dtype,
+        'doc': doc,
+        'is_db_field': is_db_field,
+        'is_subscriptable': is_subscriptable,
+    }
+    if not isidentifier(key):
+        raise ValueError(f'{key} is not a valid python identifier')
+    if is_db_field and qb_field:
+        raise ValueError('only attribute fields may be aliased')
```

Comment:
I implemented these here instead of in `QbField`, because in the case of subscripted fields, e.g. `orm.Dict.fields.dict.a`, you get a new `QbAttrField('dict.a')`, which is valid in the sense that you don't use it programmatically, but `dict.a` (or `attributes.a` post mapping conversion) does get used in the backeknd.